src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupBasedAlleles.java

```diff
@@ -31,47 +54,132 @@
+                final Optional<Map.Entry<String, Integer>> maxIns = insertionCounts.entrySet().stream().max(Comparator.comparingInt(Map.Entry::getValue));
+                if (maxIns.isPresent()
+                        && passesFilters(args, true, numOfBases, totalAltBadReads, totalAltReads, maxIns.get())) {
+
+                    indelAlleles.add(Allele.create((char)referenceContext.getBase() + maxIns.get().getKey()));
+                    final VariantContextBuilder pileupInsertion = new VariantContextBuilder("pileup", alignmentContext.getContig(), alignmentContext.getStart(), alignmentContext.getEnd(), indelAlleles);
+                    pileupSNPsList.add(pileupInsertion.make());
+                }
             }
         }
 
         return pileupSNPsList;
     }
 
+    /**
+     * Apply the filters to discovered alleles
+     * - Does it have greater than snpThreshold percentage of bases support in the pileups?
+     * - Does it have greater than pileupAbsoluteDepth number of reads supporting it?
+     * - Are the reads supporting alts at the site greater than badReadThreshold percent "good"? //TODO evaluate if this is worth doing on a per-allele basis or otherwise
+     */
+    private static boolean passesFilters(final PileupDetectionArgumentCollection args, boolean indel,  final int numOfBases, final int totalAltBadReads, final int totalAltReads, final Map.Entry<?, Integer> maxAlt) {
+        return ((float) maxAlt.getValue() / (float) numOfBases) > (indel ? args.indelThreshold : args.snpThreshold)
+                && numOfBases >= args.pileupAbsoluteDepth
+                && ((args.badReadThreshold <= 0.0) || (float) totalAltBadReads / (float)totalAltReads <= args.badReadThreshold);
+    }
+
+    /**
+     * Based on the illumina PileupDetection filtering code: We apply a number of configurable heuristics to the reads that support
+     * alt alleles that may be added and for each read evaluate if its "bad" or not by each of the heurisitcs. Currently they are:
+     * - Secondary/SA tag reads are bad
+     * - Improperly paired reads are bad
+     * - Reads with > 8% per-base edit distance to the reference are bad
+     * - Reads 2 std deviations away from the standard insert size are bad (not implemented)
+     *
+     * @param read
+     * @param referenceContext
+     * @param args
+     * @param headerForRead TODO get rid of this sam record coversion
+     * @return
+     */
+    @VisibleForTesting
+    static boolean evaluateBadRead(final GATKRead read, final ReferenceContext referenceContext, final PileupDetectionArgumentCollection args, final SAMFileHeader headerForRead) {
+        if (args.badReadProperPair && !read.isProperlyPaired()) {
+            return true;
+        }
+        if (args.badReadSecondary && (read.isSecondaryAlignment() || read.hasAttribute("SA"))) {
+            return true;
+        }
+
+
+        //TODO this conversion is really unnecessary...
```

Comment:
huh? so why is it done?

Reply:
Its done because the method `SequenceUtil.calculateSamNmTag()` works on SAMRecords and its in HTSJDK so I can't easily change it... I didn't want to reimplement it but as you can see i have been forced to thread the header down to this class as a result.